common/app/redux/entities-reducer.js

```diff
@@ -1,103 +1,94 @@
       ...state,
-      user: {
-        ...state.user,
-        [username]: {
-          ...state.user[username],
+      [username]: {
+          ...state[username],
           points
         }
-      }
-    };
-  }
-  if (action.type === types.updateUserFlag) {
-    return {
+    }),
+    [types.updateUserFlag]: (state, { payload: { username, flag } }) => ({
       ...state,
-      user: {
-        ...state.user,
-        [username]: {
-          ...state.user[username],
+      [username]: {
+          ...state[username],
           [flag]: !state.user[username][flag]
         }
-      }
-    };
-  }
-  if (action.type === types.updateUserEmail) {
-    return {
+    }),
+    [types.updateUserEmail]: (state, { payload: { username, email } }) => ({
       ...state,
-      user: {
-        ...state.user,
-        [username]: {
-          ...state.user[username],
+      [username]: {
+          ...state[username],
           email
         }
-      }
-    };
-  }
-  if (action.type === types.updateUserLang) {
-    return {
+    }),
+    [types.updateUserLang]:
+    (
+      state,
+      { payload: { username, languageTag } }
+    ) => ({
       ...state,
-      user: {
-        ...state.user,
-        [username]: {
-          ...state.user[username],
+      [username]: {
+          ...state[username],
           languageTag
         }
-      }
-    };
-  }
-
-  if (action.type === types.updateMyCurrentChallenge) {
-    return {
+    }),
+    [types.updateMyCurrentChallenge]:
+    (
+      state,
+      { payload: username, currentChallengeId }
+    ) => ({
       ...state,
-      user: {
-        ...state.user,
-        [username]: {
-          ...state.user[username],
-          currentChallengeId: action.payload.currentChallengeId
+      [username]: {
+          ...state[username],
+          currentChallengeId
         }
+    }),
+    [types.updateUserChallenge]:
+    (
+      state,
+      {
+        payload: { username, challengeInfo }
       }
-    };
-  }
-
-  if (action.type === types.updateUserChallenge) {
-    const { challengeInfo } = action.payload;
-    return {
+    ) => ({
       ...state,
-      user: {
-        ...state.user,
-        [username]: {
-          ...state.user[username],
+      [username]: {
+          ...state[username],
           challengeMap: {
-            ...state.user[username].challengeMap,
+            ...state[username].challengeMap,
             [challengeInfo.id]: challengeInfo
           }
         }
-      }
+    })
+  },
+  initialState.user
+);
+
+function metaReducer(state = initialState, action) {
+  if (action.meta && action.meta.entities) {
+    return {
+      ...state,
+      ...action.meta.entities
     };
   }
   return state;
 }
+
+export default function entitiesReducer(state, action) {
+  const newState = metaReducer(state, action);
+  const user = userReducer(state && state.user || {}, action);
+  if (newState.user !== user && Object.keys(user).length > 0) {
```

Comment:
Why are you doing the additional check for user keys? If the user ref values are not the same, that alone should signify that the user object has been updated and the overall state object should receive a  new ref value.